File: server/src/main/com/broll/gainea/server/core/battle/RollResult.java
package com.broll.gainea.server.core.battle;

import com.broll.gainea.misc.RandomUtils;
import com.broll.gainea.server.core.objects.Unit;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class RollResult {

    private List<Roll> rolls;

    public RollResult(BattleContext context, List<Unit> units) {
        this.rolls = units.stream().flatMap(unit -> unit.calcFightingPower(context).roll()
                .stream().map(value -> new Roll(unit, value))).collect(Collectors.toList());
        sort();
    }

    public void plusNumber(int number) {
        this.rolls.forEach(it -> it.value += number);
    }

    public void plusNumber(int number, int affectDiceCount) {
        Collections.shuffle(this.rolls);
        for (int i = 0; i < Math.min(affectDiceCount, rolls.size()); i++) {
            rolls.get(i).value += number;
        }
        sort();
    }

    public void max(int number) {
        this.rolls.forEach(it -> it.value = Math.max(number, it.value));
    }

    public void min(int number) {
        this.rolls.forEach(it -> it.value = Math.min(number, it.value));
    }

    public void addDice(int number) {
        this.rolls.add(new Roll(null, number));
        sort();
    }

    public void addDice() {
        addDice(RandomUtils.random(1, 6));
    }

    public void removeDice(int count) {
        for (int i = 0; i < count; i++) {
            if (rolls.size() > 1) {
                rolls.remove(RandomUtils.random(0, rolls.size() - 1));
            }
        }
        sort();
    }

    public void removeHighestDice(int count) {
        for (int i = 0; i < count; i++) {
            if (rolls.size() > 1) {
                rolls.remove(0);
            }
        }
    }

    public void removeLowestDice(int count) {
        for (int i = 0; i < count; i++) {
            if (rolls.size() > 1) {
                rolls.remove(rolls.size() - 1);
            }
        }
    }

    public int count() {
        return rolls.size();
    }

    public void sort() {
        Collections.sort(rolls, Collections.reverseOrder());
    }

    public List<Roll> getRolls() {
        return this.rolls;
    }

    public class Roll implements Comparable<Roll> {
        public Unit source;
        public int value;

        private Roll(Unit source, int value) {
            this.source = source;
            this.value = value;
        }

        public int compareTo(Roll other) {
            return Integer.compare(this.value, other.value);
        }

    }

}

File: server/src/main/com/broll/gainea/server/core/battle/RollResult.kt
package com.broll.gainea.server.core.battle

import com.broll.gainea.misc.RandomUtils
import com.broll.gainea.net.NT_Battle_Roll
import com.broll.gainea.server.core.objects.Unit

class RollResult(context: BattleContext, units: List<Unit>) {
    val rolls: MutableList<Roll>

    init {
        rolls = units.flatMap { unit ->
            val power = unit.calcFightingPower(context)
            power.roll().map { result ->
                Roll(
                        source = unit,
                        value = result,
                        plus = power.numberPlus,
                        min = power.lowestNumber,
                        max = power.highestNumber
                )
            }
        }.toMutableList()
        sort()
    }

    fun plusNumber(number: Int) {
        rolls.forEach { it.value += number }
    }

    fun plusNumber(number: Int, affectDiceCount: Int) {
        rolls.shuffle()
        for (i in 0 until Math.min(affectDiceCount, rolls.size)) {
            rolls[i].value += number
            rolls[i].plus += number
        }
        sort()
    }

    fun max(number: Int) {
        rolls.forEach {
            it.value = Math.max(number, it.value)
            it.max = number
        }
    }

    fun min(number: Int) {
        rolls.forEach {
            it.value = Math.min(number, it.value)
            it.min = number
        }
    }

    fun addDice(number: Int = RandomUtils.random(1, 6)) {
        rolls.add(Roll(null, number))
        sort()
    }

    fun removeDice(count: Int) {
        for (i in 0 until count) {
            if (rolls.size > 1) {
                rolls.removeAt(RandomUtils.random(0, rolls.size - 1))
            }
        }
        sort()
    }

    fun removeHighestDice(count: Int) {
        for (i in 0 until count) {
            if (rolls.size > 1) {
                rolls.removeAt(0)
            }
        }
    }

    fun removeLowestDice(count: Int) {
        for (i in 0 until count) {
            if (rolls.size > 1) {
                rolls.removeAt(rolls.size - 1)
            }
        }
    }

    fun count(): Int {
        return rolls.size
    }

    private fun sort() {
        rolls.sortDescending()
    }

    inner class Roll(var source: Unit?, var value: Int, var plus: Int = 0, var min: Int = 1, var max: Int = 6) : Comparable<Roll> {
        override fun compareTo(other: Roll): Int {
            return Integer.compare(value, other.value)
        }

        fun nt(): NT_Battle_Roll {
            val nt = NT_Battle_Roll()
            nt.sourceUnit = source?.id ?: -1
            nt.number = value
            nt.min = min
            nt.max = max
            nt.plus = plus
            return nt
        }
    }
}
